File: 6.栈与队列/71.简化路径.cpp
/*
 * @lc app=leetcode.cn id=71 lang=cpp
 *
 * [71] 简化路径
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    string simplifyPath(string path) {
        vector<string> stk;
        string s;
        istringstream ss(path);
        while (getline(ss, s, '/'))
        {
            if (s.empty() || s == ".") continue;
            else if (s == ".." && stk.size()) stk.pop_back();
            else if (s != "..") stk.push_back(s);
        }
        string res;
        for (auto str : stk)
        {
            res += '/';
            res += str;
        }
        return res.size() ? res : "/";
    }
};
// @lc code=end
class Solution {
    public:
        string simplifyPath(string path) {
            stack<string> stk;
            int i = 0;
            while (i < path.size())
            {
                while (i < path.size() && path[i] == '/') i ++;
                string folder;
                while (i < path.size() && path[i] != '/') folder += path[i ++];
                if (folder == ".") continue;
                else if (folder == ".." && stk.size()) stk.pop();
                else if (folder == ".." && stk.empty()) continue;
                else
                {
                    if (folder.size())
                    {
                        folder += '/';
                        stk.push(folder);
                    }
                }
            }
            stack<string> stk2;
            while (stk.size())
            {
                stk2.push(stk.top());
                stk.pop();
            }
            string res = "/";
            while (stk2.size())
            {
                res += stk2.top();
                if (stk2.size() == 1) res.pop_back();
                stk2.pop();
            }
            return res;
        }
    };

File: 面经150/面经150p1/7.栈/71.简化路径.py
#
# @lc app=leetcode.cn id=71 lang=python3
#
# [71] 简化路径
#

# @lc code=start
class Solution:
    def simplifyPath(self, path: str) -> str:
        stk = []
        n = len(path)
        res = ""
        i = 0
        while i < n:
            if path[i] == '/': i += 1; continue
            s = ""
            while i < n and path[i] != '/': s += path[i]; i += 1
            if s == "..":
                if stk: stk.pop()
                else: continue
            elif s != ".": stk.append(s)
        for s in stk: res += '/' + s
        return res if res != "" else "/"
# @lc code=end

